### PR TITLE
add metadata model

### LINKED ISSUE #30: Create a mosaic `metadata` specification
Metadata could be useful for the client but also within the API itself.

For example, when using CQL-JSON, the `bbox` is used in the tilejson endpoint https://github.com/stac-utils/titiler-pgstac/blob/0f2b5b4ba50bb3458237ab21cf9a154d7b811851/titiler/pgstac/factory.py#L265

But now with cql2, we can't do this. We would have to `parse` the cql2 query to find if an `intersect` filter was set, which seems not a great idea. On the other end if we were to specify a `bbox` (or `bounds`) in the mosaic `metadata` it would be IMO an acceptable solution.

### Metadata Spec

- **bbox** (`Tuple[float, float, float, float]`): WGS84 bounding box
- **minzoom** (`int`): Min Zoom (Note: depending on the TileMatrixSet used, the minzoom might be different, so we might use the minzoom only for the default WebMercatorQuad TMS)
- **maxzoom** (`int`): Max Zoom (Note: depending on the TileMatrixSet used, the minzoom might be different, so we might use the maxzoom only for the default WebMercatorQuad TMS)
- **assets** (`List[str]`): available assets for the mosaic
- **default-assets** (`List[str]`): `assets` to use by default when not set in the url

Note: this would be almost only used in the tilejson endpoint.

### PR DESCRIPTION
**Pull request by vincentsarago**
ref #30 

This PR adds a `SearchMetadata` model defining the `mosaic` metadata that could be provided by the user when `registering` a search request. 

#### Metadata

```js
{
    // OPTIONAL. Default: "mosaic" (No other value accepted for now). Describe the `type` of metadata. 
    "type": "mosaic",

    // OPTIONAL. Default: null.
    // The maximum extent of available map tiles. The bounds are represented in WGS:84
    // latitude and longitude values, in the order left, bottom, right, top.
    // Values may be integers or floating point numbers.
    "bounds": [ -180, -85.05112877980659, 180, 85.0511287798066 ],

    // OPTIONAL. Default: null.
    // An integer specifying the minimum zoom level.
    "minzoom": 0,

    // OPTIONAL. Default: null.
    // An integer specifying the maximum zoom level. MUST be >= minzoom.
    "maxzoom": 11,

    // OPTIONAL. Default: null. The name can contain any legal character.
    "name": "compositing",

    // OPTIONAL. Default: null. An array of available assets.
    "assets": ["image", "cog"],

    // OPTIONAL. Default: null. A set of `defaults` configuration.
    "defaults": {
        "true_color": {
            "assets": ["B4", "B3", "B2"],
            "color_formula": "Gamma RGB 3.5 Saturation 1.7 Sigmoidal RGB 15 0.35",
        },
        "ndvi": {
            "expression": "(B4-B3)/(B4+B3)",
            "rescale": "-1,1",
            "colormap_name": "viridis"
        }
    }
}
```

### CODE REVIEW

File: titiler/pgstac/factory.py [hunk 1]
```diff
@@ -31,15 +30,16 @@ def PathParams(searchid: str = Path(..., description="Search Id")) -> str:
...
     """Search parameters."""
-    search = body.json(
+    search = body.dict(
         exclude_none=True,
         exclude={"metadata"},
         by_alias=True,
     )
-    metadata = body.metadata or {}
-    return search, metadata
+    return model.PgSTACSearch(**search), body.metadata
```
Comment (vincentsarago) [inline titiler/pgstac/factory.py:42]: FYI: @drnextgis

File: titiler/pgstac/model.py [hunk 2]
```diff
@@ -129,3 +173,9 @@ def validate_bbox(cls, v: BBox):
                 raise ValueError("Bounding box must be within (-180, -90, 180, 90)")
 
         return v
+
+
+class RegisterMosaic(PgSTACSearch):
+    """Model of /register endpoint input."""
+
+    metadata: Metadata = Field(default_factory=Metadata)
```
Comment (vincentsarago) [inline titiler/pgstac/model.py:181]: Metadata will always be set when registering a `search` with the `RegisterMosaic`. default metadata should be `{"type": "mosaic"}`

File: titiler/pgstac/factory.py (hunk 1 above)
Comment (drnextgis) [reply titiler/pgstac/factory.py:42]: Awesome! As soon as it is merged, I will update the example in the documentation.

**Comment by drnextgis** [comment]
Looks good to me. Just one question: what is the purpose of having `SearchType.search`?

**Comment by vincentsarago** [comment]
@drnextgis 

> what is the purpose of having SearchType.search?

I guess my goal is to quickly differentiate search requests `registered` through titiler-pgstac and other search entries. This way when we do #40 we can only filter on `type=="mosaic"`. the type `SearchType.search ` will be set by default for `search` entries that doesn't have the type `SearchType.mosaic` (not used for now, but maybe later)

**Comment by vincentsarago** [comment]
FYI: I'm going to merge this to Master, but anyone interested can still make comments over https://github.com/stac-utils/titiler-pgstac/issues/30

File: titiler/pgstac/factory.py (hunk 1 above)
Comment (vincentsarago) [reply titiler/pgstac/factory.py:43]: https://github.com/stac-utils/titiler-pgstac/commit/d3eecdcbcf23aef59fd240914441ef27b3e885ee#diff-0f0c40de44a0841ed5180f16d66212d1cece2a7d4400789007c2f73bf8432622L39 I updated the docs 🙏

**Pull request merged**